src/vs/workbench/parts/terminal/node/terminalPanel.ts

```diff
@@ -0,0 +1,145 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import termJs = require('term.js');
+import fs = require('fs');
+import {fork, Terminal} from 'pty.js';
+import platform = require('vs/base/common/platform');
+import {TPromise} from 'vs/base/common/winjs.base';
+import {Builder, Dimension} from 'vs/base/browser/builder';
+import {IConfigurationService} from 'vs/platform/configuration/common/configuration';
+import {ITelemetryService} from 'vs/platform/telemetry/common/telemetry';
+import {IWorkspaceContextService} from 'vs/platform/workspace/common/workspace';
+import {ITerminalConfiguration, TERMINAL_PANEL_ID} from 'vs/workbench/parts/terminal/common/terminal';
+import {Panel} from 'vs/workbench/browser/panel';
+import {ScrollableElement} from 'vs/base/browser/ui/scrollbar/scrollableElementImpl';
+import {DomNodeScrollable} from 'vs/base/browser/ui/scrollbar/domNodeScrollable';
+
+const TERMINAL_CHAR_WIDTH = 8;
+const TERMINAL_CHAR_HEIGHT = 18;
+
+export class TerminalPanel extends Panel {
+
+	private ptyProcess: Terminal;
+	private parentDomElement: HTMLElement;
+	private terminal;
+	private terminalDomElement: HTMLDivElement;
+
+	constructor(
+		@IConfigurationService private configurationService: IConfigurationService,
+		@ITelemetryService telemetryService: ITelemetryService,
+		@IWorkspaceContextService private contextService: IWorkspaceContextService
+	) {
+		super(TERMINAL_PANEL_ID, telemetryService);
+	}
+
+	public layout(dimension: Dimension): void {
+		let cols = Math.floor(this.parentDomElement.offsetWidth / TERMINAL_CHAR_WIDTH);
+		let rows = Math.floor(this.parentDomElement.offsetHeight / TERMINAL_CHAR_HEIGHT);
+		this.terminal.resize(cols, rows);
+		this.ptyProcess.resize(cols, rows);
+	}
+
+	public create(parent: Builder): TPromise<void> {
+		super.create(parent);
+
+		this.parentDomElement = parent.getHTMLElement();
+		this.createTerminal();
+
+		return TPromise.as(null);
+	}
+
+	private createTerminal(): void {
+		this.parentDomElement.innerHTML = '';
+		this.ptyProcess = fork(this.getShell(), [], {
+			name: fs.existsSync('/usr/share/terminfo/x/xterm-256color') ? 'xterm-256color' : 'xterm',
```

Comment:
⚡ evil sync io inside the main thread. use callbacks/promises esp given your outer caller (`create`) allows you to do that